pkg/engine/runtime/runtime.go

```diff
@@ -9,25 +9,80 @@ import (
-	// Apply means modify this resource to the desired state described in the request,
-	// and it will turn into creating or updating a resource in most scenarios.
+	// Apply means modify this Resource to the desired state described in the request,
+	// and it will turn into creating or updating a Resource in most scenarios.
 	// If the infrastructure runtime already provides an Apply method that conform to this method's semantics meaning,
 	// like the Kubernetes Runtime, you can directly invoke this method without any conversion.
-	// PlanState and priorState are given in this method for the runtime which would make a
+	// PlanResource and priorState are given in this method for the runtime which would make a
 	// three-way-merge (planState,priorState and live state) when implementing this interface
-	Apply(ctx context.Context, priorState, planState *models.Resource) (*models.Resource, status.Status)
+	Apply(ctx context.Context, applyStruct *ApplyRequest) *ApplyResponse
 
-	// Read the latest state of this resource
-	Read(ctx context.Context, resourceState *models.Resource) (*models.Resource, status.Status)
+	// Read the latest state of this Resource
+	Read(ctx context.Context, request *ReadRequest) *ReadResponse
 
-	// Delete this resource in the actual infrastructure and return success if this resource is not exist
-	Delete(ctx context.Context, resourceState *models.Resource) status.Status
+	// Delete this Resource in the actual infrastructure and return success if this Resource is not exist
+	Delete(ctx context.Context, request *DeleteRequest) *DeleteResponse
 
-	// Watch the latest state or event of this resource.
+	// Watch the latest state or event of this Resource.
 	// This is an optional method for the Runtime to implement,
-	// but it will be very helpful for us to know what is happening when applying this resource
-	Watch(ctx context.Context, resourceState *models.Resource) (*models.Resource, status.Status)
+	// but it will be very helpful for us to know what is happening when applying this Resource
+	Watch(ctx context.Context, request *WatchRequest) *WatchResponse
+}
+
+type ApplyRequest struct {
+	// PriorResource is the last applied resource saved in state storage
+	PriorResource *models.Resource
+
+	// PlanResource is the resource we want to apply in this request
+	PlanResource *models.Resource
+
+	// DryRun means this a dry-run request and will not make any changes in actual infra
+	DryRun bool
+}
+
+type ApplyResponse struct {
+	// Resource is the result returned by Runtime
+	Resource *models.Resource
+
+	// Status contains messages will show to users
+	Status status.Status
+}
+
+type ReadRequest struct {
+	// Resource represents the resource we want to read from the actual infra
+	Resource *models.Resource
+}
+
+type ReadResponse struct {
+	// Resource is the result read from the actual infra
+	Resource *models.Resource
+
+	// Status contains messages will show to users
+	Status status.Status
+}
+
+type DeleteRequest struct {
```

Comment:
should DeleteRequest support dry-run?

Reply:
Dry-run in apply request is designed for 3-way diff. Dry-run in delete is not required at this moment